gemoji/src/main/kotlin/io/sweers/catchup/gemoji/EmojiMarkdownConverter.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2017 Zac Sweers
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.sweers.catchup.gemoji
+
+/**
+ * Converts a markdown emoji alias (eg, ":smile:") into an android render-able emoji.
+ */
+interface EmojiMarkdownConverter {
+  fun convert(alias: String): String?
+}
+
+class GemojiEmojiMarkdownConverter(private val gemojiDao: GemojiDao): EmojiMarkdownConverter {
+  override fun convert(alias: String): String? {
+    if (alias.startsWith(':') && alias.endsWith(":")) {
+      return gemojiDao.getEmoji(alias.substring(1, alias.lastIndex))
+    } else {
+      return null
+    }
+  }
+}
+
+/**
+ * Returns a [String] that replaces occurrences of markdown emojis with android render-able emojis.
+ */
+fun replaceMarkdownEmojis(markdown: String, converter: EmojiMarkdownConverter): String {
+  val sb = StringBuilder(markdown.length)
+  var potentialAliasStart: Int? = null
```

Comment:
nit: since we're bangbanging it below already and not getting the nullability benefit, should we just make this a regular Int and use -1 as a NO_INDEX state?